Implemented the Bookings UI

MilesDev91 opened the pull request:
### What this PR does:
- Provides the skeleton to the Bookings page. There is a section for the bookings and a calendar we can pick dates from.
- Bookings are displayed on bottom on mobile and the left side on desktop. Calendar is on the top or the right.
- Calendar displays all upcoming non-declined bookings.

### Screenshots / Videos:
- Desktop:
![image](https://user-images.githubusercontent.com/34494229/149601115-2d9697d5-5923-42fc-a488-256d79f740e7.png)
- Mobile:
![image](https://user-images.githubusercontent.com/34494229/149601135-3d18d7cb-e62c-4a6d-9f84-c9df3f8a0da3.png)
![image](https://user-images.githubusercontent.com/34494229/149601156-1b7f3cd8-48ec-4ab3-965c-e6d68861894d.png)


### Any information needed to test this feature:
- Make sure you are logged in as any user, then navigate to /bookings.
- Click on the calendar dates and navigation features as well.

### Any issues with the current functionality:
- Bookings are mock data currently.
- We don't currently handle empty bookings. This will be easier to implement once we know the exact way we retrieve the information, which is during integration.

closes #19 

### Review thoughts:
- I wanted to use react-calendar because it had decent functionality already built in and thought it would save some time. The styling side of it was awkward, and I didn't find elegant solutions to some of the problems (like using padding for backgrounds like normal).
- I added mock information and the means to display it. The UI is much more complete now. I also feel a bit better about the calendar (the active styling is the same across all screen sizes for instance).

MilesDev91 committed:
Added bookings page and end of line

Added the bookings page to the file structure and App.tsx routes.

Added some settings for local mongo database connection to the .env and www files.

Changed file endings to LF and added a .gitattributes file to ensure they don't get changed back for users who have Windows git configurations.

MilesDev91 committed:
Reversed changes to the environment files

Added a .env file at the root of server with settings for my own local connection so they don't interfere with other developer's needs.

MilesDev91 committed:
Started Bookings styling

Started styling the Bookings page to match the mocks.

Added react-calendar as our calendar of choice. Began styling the calendar.

MilesDev91 committed:
Styling the calendar

Adjusted some styling for spacing and flex directions.

Styled the calendar to match the mocks as closely as I can.

MilesDev91 committed:
Created Calendar component

Separated the react calendar into its own component so we can reuse the styling modifications throughout the application.

MilesDev91 committed:
Update useStyles.ts

Added a comment for clarification

ugur-akin commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -0,0 +1,23 @@
+
+  return (
+    <PageContainer>
+      <Grid container classes={{ root: classes.root }} spacing={{ md: 5 }}>
```
You can apply this class to the root element directly via `className` as well 👍🏼

ugur-akin commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -0,0 +1,23 @@
+          <Calendar />
+        </Grid>
+        <Grid xs={10} md={3} direction="column" item container className={classes.upcomingBookings}>
+          <Box>Your next booking:</Box>
+          <Box>Current bookings:</Box>
```
We should implement the UI for booking cards and interacting with them as they are a part of this ticket 👍🏼 I'd recommend using some mock booking data to fill out the respective parts of the UI dynamically still. 
<img width="292" alt="Manage Booking" src="https://user-images.githubusercontent.com/70963670/149496474-a6dc80c8-1055-46b6-ad14-fafa9319845a.png">

MilesDev91 replied:
Will do.

ugur-akin commented on client/src/pages/Bookings/useStyles.ts:
```diff
@@ -0,0 +1,27 @@
+    '@media(min-width: 600px)': {
+      order: -1,
+    },
+    '& .MuiBox-root': {
+      boxShadow:
+        '0px 0px 1.9px rgba(0, 0, 0, 0.007),0px 0px 4.9px rgba(0, 0, 0, 0.014),0px 0px 9.9px rgba(0, 0, 0, 0.021),0px 0px 20.4px rgba(0, 0, 0, 0.031),0px 0px 56px rgba(0, 0, 0, 0.05)',
+      marginBottom: theme.spacing(2),
+      padding: theme.spacing(3),
+      textTransform: 'uppercase',
+      fontSize: '.7rem',
+      fontWeight: 'bold',
+    },
+    '& .MuiGrid-root': {
+      flexDirection: 'column',
+    },
```
Ideally, we add these styles directly to the components instead of targeting nested elements with CSS. You can also apply some of the styles here directly to the components (e.g. both `Box` and `Grid` components accept all [system properties](https://mui.com/system/properties/)). Use your discretion to decide what styles are good to see explicitly and what's better to tuck away in order to keep the code more readable.

MilesDev91 replied:
Oh I did not realize that. I will certainly use those more often. Thank you!

ugur-akin commented on client/src/pages/Bookings/useStyles.ts:
```diff
@@ -0,0 +1,27 @@
+import { makeStyles } from '@mui/styles';
+
+const useStyles = makeStyles((theme: Theme) => ({
+  root: {
+    justifyContent: 'center',
+  },
```
Same goes for this 👍🏼

ugur-akin reviewed (changes requested):
Hey Miles, great work with this so far! 💯  You mentioned you were wondering if you should use mock data to build the UI for bookings, yes we should include that part of the UI in this pull 👍🏼 I also highlighted some thoughts about styling components with mui system, feel free to take a look and ask any questions or share any thoughts! 

Thanks for well presenting the details of your PR in your title and description! 🙏🏼 

Cheers 🎉

MilesDev91 committed:
Implementing mock bookings

Added a booking interface to the interface folder.

Created BookingCard component that handles the display of the given booking information.

Continuing to style and implement information in the provided bookings

MilesDev91 committed:
Finished styling bookings and calendar

Formatted and displayed bookings, styled appropriately. Bookings are sorted out by their start dates in the Bookings page component, then use several BookingCard components to format and display the information for the specified booking.

Implemented dynamic styling for the calendar tiles based on upcoming bookings. Declined bookings do not show on the calendar. This specific styling proved to be more elegant than the previous way to highlight, so adjusted the active styling elsewhere as well.

Adjusted props in several components as needed.

rajivtitus commented on client/src/components/BookingCard/BookingCard.tsx:
```diff
@@ -0,0 +1,70 @@
+}
+
+export default function BookingCard({ booking, isNextBooking }: Props): JSX.Element {
+  const date = () => {
```
Just a note: These type of functions can be moved to a `utils` folder, in case we need to use the same date format in other components. However, for our project I don't think this is the case, so no action needed 😄

rajivtitus commented on client/src/components/BookingCard/BookingCard.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  // The top line needs to be structured differently depending on if it is in the next-booking box or in the bottom box
+  const topLine = () => {
+    if (isNextBooking) {
```
Nit: Using a ternary here would make the code look a bit more cleaner :)

`{isNextBooking ? ( <JSX /> ) : ( <JSX /> )}`

rajivtitus commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -0,0 +1,127 @@
+        </Grid>
+        <Grid xs={10} md={3} direction="column" item container className={classes.bookings}>
+          <Box sx={{ padding: '24px', marginBottom: '16px' }}>
+            <div className={classes.bookingSectionLabel}>
```
Let's resort to using MUI `Box` and `Typography` components instead of `div` and `p` to go with MUI semantics

MilesDev91 replied:
Ah keep reverting to established habits

rajivtitus reviewed (commented):
- Good PR description
- Nice touch with documenting the code where necessary
- UI look great 💯 
Excellent work Miles, just some minor comments to be addressed :)

rajivtitus commented on client/src/pages/Bookings/Bookings.tsx:
```diff
@@ -0,0 +1,127 @@
+  const pastBookings = getPastBookings(mockBookings);
+
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const nextBooking: Booking = mockBookings.shift()!;
```
After reading a bit more about the 'no-non-null-assertion', I'm leaning towards another approach instead of us assuming objects exist (as using non-null assertion cancels the benefits of the strict null checking mode) :)

MilesDev91 replied:
I think I have an idea of what to do

MilesDev91 committed:
Handled empty bookings and better styling

Cleaned up some code in the booking card.

Handled the case where user has either no past bookings, no current bookings, or any combination of the two. Will display the next booking box at a minimum with content or a message.

Changed some HTML to use MUI elements instead of <p> or <div> elements.

Changed targeted CSS to allow for more granularity down the element tree.

MilesDev91 commented:
Let me know if I missed anything or if the Bookings.tsx file needs organized/commented better now that I added cases for empty booking schedules.

rajivtitus reviewed (approved):
LGTM 👍🏽

MilesDev91 committed:
Merge branch 'main' into bookings-ui
